### PR TITLE
[mono] Optimize mono_metadata_decode_row_col_raw

### PR DESCRIPTION
**Pull request by kg**
The loop in this function is somewhat expensive, and in some wasm startup profiles I did we spent around 1% of cpu time inside it. From some quick tests this seems to make it a little bit faster, at the price of 9 bytes per MonoTableInfo (~500 bytes per assembly), which I think is worth it.

### CODE REVIEW

File: src/mono/mono/metadata/metadata.c [hunk 1]
```diff
@@ -1003,6 +1003,21 @@ mono_metadata_table_bounds_check_slow (MonoImage *image, int table_index, int to
         return mono_metadata_update_table_bounds_check (image, table_index, token_index);
 }
 
+static void
+mono_metadata_compute_column_offsets (guint32 bitfield, guint8 column_offsets[9])
+{
```
Comment (vargaz) [inline src/mono/mono/metadata/metadata.c:1008]: This doesn't need a mono_metadata prefix.

File: src/mono/mono/metadata/metadata-internals.h [hunk 2]
```diff
@@ -234,6 +234,12 @@ struct _MonoTableInfo {
 	 * we only need 4, but 8 is aligned no shift required.
 	 */
 	guint32   size_bitfield;
+
+	/*
+	 * optimize out the loop in mono_metadata_decode_row_col_raw.
+	 * 4 * 9 easily fits in a uint8
+	 */
+	guint8    column_offsets[9];
```
Comment (vargaz) [inline src/mono/mono/metadata/metadata-internals.h:242]: Make 9 a named constant ?

**Review by vargaz — APPROVED** [review]

File: src/mono/mono/metadata/metadata.c [hunk 3]
```diff
@@ -1003,6 +1003,21 @@ mono_metadata_table_bounds_check_slow (MonoImage *image, int table_index, int to
         return mono_metadata_update_table_bounds_check (image, table_index, token_index);
 }
 
+static void
+compute_column_offsets (guint32 bitfield, guint8 column_offsets[MONO_TABLE_INFO_MAX_COLUMNS])
```
Comment (lambdageek) [inline src/mono/mono/metadata/metadata.c:1007]: nit: we don't usually use fixed length array types in function arguments in mono. we usually just use a pointer type. probably doesn't matter.  just stands out a bit.

File: src/mono/mono/metadata/metadata.c [hunk 4]
```diff
@@ -1003,6 +1003,21 @@ mono_metadata_table_bounds_check_slow (MonoImage *image, int table_index, int to
...
+	int offset = 0, c = mono_metadata_table_count (bitfield);
+	for (int i = 0; i < MONO_TABLE_INFO_MAX_COLUMNS; i++) {
+		if (i >= c) {
+			column_offsets[i] = 0xFFu;
+		} else {
+			int size = mono_metadata_table_size (bitfield, i);
+			column_offsets[i] = (guint8)offset;
+			offset += size;
+		}
+	}
```
Comment (lambdageek) [inline src/mono/mono/metadata/metadata.c:1018]: probably doesn't matter, but why not `memset` and then overwrite just the fields from 0 to c-1?  

Also why do we need to initialize the out of range fields to a non-zero value?

File: src/mono/mono/metadata/metadata.c [hunk 5]
```diff
@@ -1470,18 +1486,18 @@ mono_metadata_decode_row_col_raw (const MonoTableInfo *t, int idx, guint col)
...
 	data = t->base + idx * t->row_size;
 
-	n = mono_metadata_table_size (bitfield, 0);
-	for (guint i = 0; i < col; ++i) {
-		data += n;
-		n = mono_metadata_table_size (bitfield, i + 1);
-	}
+	n = mono_metadata_table_size (bitfield, col);
+	column_offset = t->column_offsets [col];
+	g_assert (column_offset < 0xFFu);
```
Comment (lambdageek) [inline src/mono/mono/metadata/metadata.c:1499]: Does this assert anything different than what line 1494 asserts: `g_assert (col < mono_metadata_table_count (bitfield));` ?

**Review by lambdageek — COMMENTED** [review]
lgtm, but we could do less, I think?

File: src/mono/mono/metadata/metadata.c (hunk 4 above)
Comment (kg) [reply src/mono/mono/metadata/metadata.c:1018]: 0 is a valid column offset, so the out of range ones are an invalid offset which we can assert against
could do memset, yeah. will change it if you want

File: src/mono/mono/metadata/metadata.c (hunk 5 above)
Comment (kg) [reply src/mono/mono/metadata/metadata.c:1499]: it's effectively the same assertion, it's just validating the contents of the lookup table in addition to the input parameter

**Review by lambdageek — APPROVED** [review]

**Pull request merged**